lib/spack/spack/solver/asp.py

```diff
@@ -124,14 +124,18 @@ def __str__(self):
 
 
 @contextmanager
-def spec_with_name(spec, name):
-    """Context manager to temporarily set the name of a spec"""
-    old_name = spec.name
-    spec.name = name
+def spec_with_name(*spec_name_tuples: Tuple[Optional["spack.spec.Spec"], Optional[str]]):
+    """Context manager to temporarily set names on specs"""
+    old_names = [spec.name if spec else None for spec, _ in spec_name_tuples]
+    for spec, name in spec_name_tuples:
+        if spec:
+            spec.name = name
     try:
-        yield spec
+        yield
     finally:
-        spec.name = old_name
+        for (spec, _), old_name in zip(spec_name_tuples, old_names):
+            if spec:
+                spec.name = old_name
```

Comment:
yeah no, it's not better 😅 . I renamed it to `named_spec` to make it a bit shorter, and we don't have to nest them. Python supports `with a(), b(), c():` since 3.1.